multi/stm32l4-multi/libmulti/libdma.c

```diff
@@ -11,205 +11,452 @@
+	handle_t interruptCond = (cond == NULL) ? dma_common[dma].cond : *cond;
 
-	if (tcie)
-		/* Enable Transfer Complete interrupt, enable channel */
-		*(channel_base + ccr) |= (1 << 1) | 0x1;
-	else
-		/* Enable channel */
-		*(channel_base + ccr) |= 0x1;
+	interrupt(irqnum, libdma_irqHandler, (void *)dma, interruptCond, NULL);
 
-	dataBarier();
+	*(channelBase + ccr) = ((priority & 0x3) << 12) | ((msize & 0x3) << 10) | ((psize & 0x3) << 8) |
+		((minc & 0x1) << 7) | ((pinc & 0x1) << 6) | ((dir & 0x1) << 4);
+	*(channelBase + cpar) = (unsigned int)paddr;
+	tmp = *(dma_common[dma].base + cselr) & ~(0xF << channel * 4);
+	*(dma_common[dma].base + cselr) = tmp | ((unsigned int)reqmap << channel * 4);
 }
 
 
-static void _unprepare_transfer(volatile unsigned int *channel_base)
+int libdma_configurePeripheral(const struct libdma_per *per, int dir, int priority, void *paddr, int msize, int psize, int minc, int pinc, handle_t *cond)
 {
-	dataBarier();
-	/* Disable Transfer Complete interrupt, disable channel */
-	*(channel_base + ccr) &= ~((1 << 1) | 0x1);
-	dataBarier();
+	int dma = per->dma, channel = per->channel[dir];
+	char reqmap = per->reqmap;
+
+	libdma_configureChannel(dma, channel, dir, priority, paddr, msize, psize, minc, pinc, reqmap, cond);
+
+	return 0;
 }
 
 
-static int _has_channel_finished(const void *maddr, volatile unsigned int *channel_base)
+static int libdma_hasChannelFinished(volatile unsigned int *channelBase)
 {
-	return (maddr != NULL && *(channel_base + cndtr) > 0);
+	return (*(channelBase + cndtr) > 0) ? 1 : 0;
 }
 
 
-static int _transfer(int dma, int rx_channel, int tx_channel, void *rx_maddr, const void *tx_maddr, size_t len)
+static int libdma_transferHelper(int dma, int rxChannel, int txChannel, void *rxMAddr, const void *txMAddr, size_t len)
 {
 	/* Empirically chosen value to avoid mutex+cond overhead for short transactions */
 	int use_interrupts = len > 24;
+	int interrupts_flags = (use_interrupts == 0) ? 0 : DMA_TCIE_FLAG;
+	volatile int rxDone = 0;
 
-	volatile unsigned int *rx_channel_base = dma_common[dma].base + 2 + 5 * rx_channel;
-	volatile unsigned int *tx_channel_base = dma_common[dma].base + 2 + 5 * tx_channel;
+	volatile unsigned int *rxChannel_base = dma_common[dma].base + 2 + (5 * rxChannel);
+	volatile unsigned int *txChannel_base = dma_common[dma].base + 2 + (5 * txChannel);
+
+	/* Only one request may be issued on one channel at one time. */
+	if ((dma_transfers[dma][rxChannel].type != dma_transferNull) ||
+		(dma_transfers[dma][txChannel].type != dma_transferNull)) {
+		return -EINVAL;
+	}
 
-	if (rx_maddr != NULL)
-		_prepare_transfer(rx_channel_base, rx_maddr, len, use_interrupts);
+	dma_transfers[dma][rxChannel].type = dma_transferOnce;
+	dma_transfers[dma][rxChannel].once.doneFlag = &rxDone;
+	dma_transfers[dma][rxChannel].once.channelBase = rxChannel_base;
+	libdma_prepareTransfer(rxChannel_base, rxMAddr, len, interrupts_flags);
```

Comment:
There is probably no need to configure RX channel when we have a TX only transfer (when rxMAddr == NULL).